core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -175,6 +175,33 @@ public void afterElementUnassigned(ListVariableDescriptor<Solution_> variableDes
 
     public void beforeListVariableChanged(ListVariableDescriptor<Solution_> variableDescriptor, Object entity, int fromIndex,
             int toIndex) {
+        // Pinning is implemented in generic moves, but custom moves need to take it into account as well.
+        // This fail-fast exists to detect situations where pinned things are being moved, in case of user error.
+        var entityDescriptor = variableDescriptor.getEntityDescriptor();
+        var pinningStatus = entityDescriptor.extractPinningStatus(scoreDirector, entity);
+        if (pinningStatus.hasPin()) {
+            if (pinningStatus.entireEntityPinned()) {
+                throw new IllegalStateException("""
+                        Attempting to change list variable (%s) on an entity (%s) which is fully pinned.
+                        This is most likely a bug in a move.
+                        Maybe you are using an improperly implemented custom move?"""
+                        .formatted(variableDescriptor, entity));
+            }
+            int firstUnpinnedIndex = pinningStatus.firstUnpinnedIndex();
+            if (fromIndex < firstUnpinnedIndex || toIndex < firstUnpinnedIndex) {
+                throw new IllegalStateException(
+                        """
+                                Attempting to change list variable (%s) on an entity (%s) in range [%d, %d), but the variable's first unpinned index is (%d).
+                                This is most likely a bug in a move.
+                                Maybe you are using an improperly implemented custom move?"""
+                                .formatted(variableDescriptor, entity, fromIndex, toIndex, firstUnpinnedIndex));
+            }
+        }
```

Comment:
Very useful, but needs benchmarking to ensure it doesn't slow the hot path too much.